### PR TITLE
#28 checks if produces application/json in the problem-json rule

### PR DESCRIPTION
**Pull request by cogitor**
A possible solution for #28, when validating the Problems, check if the Produces set contains application/json, both on the top level and in operations

### CODE REVIEW

File: server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt [hunk 1]
```diff
@@ -2,12 +2,7 @@ package de.zalando.zally.rule
 
 import de.zalando.zally.violation.Violation
 import de.zalando.zally.violation.ViolationType
-import io.swagger.models.ComposedModel
-import io.swagger.models.HttpMethod
-import io.swagger.models.Model
-import io.swagger.models.RefModel
-import io.swagger.models.Response
-import io.swagger.models.Swagger
+import io.swagger.models.*
```
Comment (netme) [inline server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt]: I bet `ktlint` will complain here :-)

File: server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt [hunk 2]
```diff
@@ -56,7 +54,23 @@ class UseProblemJsonRule : AbstractRule() {
         }
     }
 
+    private fun producesJson(swagger: Swagger, operation: Operation): Boolean {
```
Comment (netme) [inline server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt]: Optional: you can also do:

```kotlin
private fun producesJson(swagger: Swagger, operation: Operation): Boolean = 
    if (operation.produces == null || operation.produces.isEmpty()) {
    ...
```

File: server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt [hunk 3]
```diff
@@ -56,7 +54,23 @@ class UseProblemJsonRule : AbstractRule() {
...
+        if (operation.produces == null || operation.produces.isEmpty()) {
```
Comment (netme) [inline server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt]: Sounds like not Kotlin :-) 
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/is-null-or-empty.html

File: server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt [hunk 4]
```diff
@@ -56,7 +54,23 @@ class UseProblemJsonRule : AbstractRule() {
...
+            return swagger.produces != null && swagger.producesContainsJson()
```
Comment (netme) [inline server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt]: This `null` check can be also removed using something like:
```kotlin
swagger.produces.orEmpty().producesContainsJson()
```

File: server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt [hunk 5]
```diff
@@ -56,7 +54,23 @@ class UseProblemJsonRule : AbstractRule() {
...
+        }
+    }
+
+    // support for application/json also with set charset e.g. "application/json; charset=utf-8"
+    private fun Swagger.producesContainsJson() =
+        produces.any { it.startsWith("application/json") }
+
+    // support for application/json also with set charset e.g. "application/json; charset=utf-8"
+    private fun Operation.producesContainsJson() =
+        produces.any { it.startsWith("application/json") }
```
Comment (netme) [inline server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt]: If `produces` is a string here and on line 67, why not to declare that on `String.isApplicationJson` level?

Comment (netme) [inline server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt]: And what is API produces `application/xml`?

File: server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt (hunk 1 above)
Comment (cogitor) [inline server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt]: that would be funny cause this was IntelliJ changing automatically :D

File: server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt (hunk 2 above)
Comment (cogitor) [inline server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt]: oh that's nice

File: server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt (hunk 1 above)
Comment (netme) [inline server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt]: Yeah, intellij is bad with default settings, same for Java

File: server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt (hunk 3 above)
Comment (cogitor) [inline server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt]: true.. but the isNullOrEmpty exists only for strings? I would need it for Lists.. now I could write it of course `private fun List.isNullOrEmpty() ...` :)

File: server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt (hunk 4 above)
Comment (cogitor) [inline server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt]: that's much nicer!

File: server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt (hunk 5 above)
Comment (cogitor) [inline server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt]: ah, well, it's actually a List.. and we're checking if any of the strings in the list is application/json or a derivative of the same ..

File: server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt (hunk 3 above)
Comment (netme) [inline server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt]: Maybe `list.orEmpty().isEmpty()` will help you here.

File: server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt (hunk 5 above)
Comment (netme) [inline server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt]: Ah, sorry, overlooked it. Then at least we can introduce this method once on the `List` object.

File: server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt (hunk 3 above)
Comment (cogitor) [inline server/src/main/java/de/zalando/zally/rule/UseProblemJsonRule.kt]: yes, that seems very readable and avoids the null check 👍

**Comment by cogitor** [comment]
@netme thanks for the feedback, I've some made changes based on it now

**Comment by netme** [comment]
👍

**Comment by netme** [comment]
👍

**Pull request merged**